### ISSUE TITLE
I think discriminator_loss calculation is flipped?

### ISSUE THREAD

**Issue by bluerand4** [UW-Madison-Lee-Lab/SFT-PG#4]
I have few questions about the code.

1. in finetune.py, the line 338 

output = f(rescale_train(torch.cat((img.detach(),x0.detach()),0)))
            
d_loss = output[:x0.shape[0]].mean()-output[x0.shape[0]:].mean()

this means I will take the 'output of first half' - 'output of second half'
first half is the 'real image' = img.
second half is the 'fake image' = x0

So this will mean like  loss = loss_real - loss_fake

but in the paper, equation 6 is saying loss = loss_fake - loss_real.

It is flipped. Am I right?

2. And also, I don't understand why you want to train 'net' in a way that it will output the 'amount of noise' that looks different from what you will see from DDIM sampled 'amount of noise'.

In line 398 from finetune.py, you make adv as constant so in order to minimize your loss, you can only change 'log_prob' to make it more negative. In order for the log_prob to be more negative, you want your prob of seeing 'ns' very low like prob = 0.000001.
But if you just try to encourage your 'net' to output the 'amount of noise' in a way that it is unlikely to see your 'ns', there are millions of ways it can output wrong. I don't think it will guarantee it will output 'amount of noise' in a way that it actually look like 'real' , not 'fake'.

Is my understanding correct?

**Comment by bluerand4** [comment]
> Hi, thanks for your question.
> 
> 1. There is no error here. In eq. 7, we use the formula in eq. 6 as sup, which mean it is a maximum term. To use torch minimization we need to flip it.
> 2. Our loss is a minimax loss, which means we want to minimize the maximum discrepancy given by the discriminator, so we minimize the IPM between fake and real, where IPM is given by the discriminator.

Thank you for your answer. I see. I get it mathematically. But l need to take some time to fully and intuitively understand the entire concept.
